File: src/models/pinCodeSession.js
/**
 * PIN码登录会话管理 - 异步 I/O 版本
 */
const fs = require('fs').promises;
const path = require('path');

// PIN码有效期 (5分钟)
const PIN_EXPIRY = 5 * 60 * 1000;

// 数据文件路径
const PIN_FILE = path.join(__dirname, '../../data/pinSessions.json');

// 状态枚举
const PINStatus = {
  PENDING: 'pending',
  USED: 'used',
  EXPIRED: 'expired'
};

// 内存缓存
let pinCache = null;
let cacheTime = 0;
const CACHE_TTL = 2000; // 2秒缓存

/**
 * 确保数据目录存在
 */
async function ensureDataDir() {
  const dataDir = path.dirname(PIN_FILE);
  await fs.mkdir(dataDir, { recursive: true }).catch(err => {
    if (err.code !== 'EEXIST') throw err;
  });
}

/**
 * 读取所有PIN码会话（带缓存）
 */
async function readPinSessions() {
  // 缓存命中
  if (pinCache !== null && (Date.now() - cacheTime) < CACHE_TTL) {
    return pinCache;
  }
  
  try {
    const data = await fs.readFile(PIN_FILE, 'utf8');
    pinCache = JSON.parse(data);
    cacheTime = Date.now();
    return pinCache;
  } catch (err) {
    if (err.code === 'ENOENT') {
      pinCache = {};
      cacheTime = Date.now();
      return pinCache;
    }
    throw err;
  }
}

/**
 * 保存所有PIN码会话
 */
async function writePinSessions(sessions) {
  await ensureDataDir();
  await fs.writeFile(PIN_FILE, JSON.stringify(sessions, null, 2), 'utf8');
  // 更新缓存
  pinCache = sessions;
  cacheTime = Date.now();
}

/**
 * 生成4位数字PIN码
 */
function generatePinCode() {
  return Math.floor(1000 + Math.random() * 9000).toString();
}

/**
 * 创建PIN码会话
 * @param {string} username 用户名
 * @returns {Promise<object>} { pin, expiresAt }
 */
async function createPinSession(username) {
  const sessions = await readPinSessions();
  
  // 删除该用户之前所有未使用的PIN码
  for (const [pin, session] of Object.entries(sessions)) {
    if (session.username === username && session.status === PINStatus.PENDING) {
      delete sessions[pin];
    }
  }
  
  // 生成唯一PIN码
  let pin = generatePinCode();
  while (sessions[pin]) {
    pin = generatePinCode();
  }
  
  const pinSession = {
    pin,
    username,
    status: PINStatus.PENDING,
    createdAt: Date.now(),
    expiresAt: Date.now() + PIN_EXPIRY
  };
  
  sessions[pin] = pinSession;
  await writePinSessions(sessions);
  
  return { pin, expiresAt: pinSession.expiresAt };
}

/**
 * 获取PIN码会话
 * @param {string} pin 
 * @returns {Promise<object|null>}
 */
async function getPinSession(pin) {
  if (!pin) return null;
  
  const sessions = await readPinSessions();
  const pinSession = sessions[pin];
  
  if (!pinSession) return null;
  
  // 检查是否过期
  if (Date.now() > pinSession.expiresAt) {
    pinSession.status = PINStatus.EXPIRED;
    sessions[pin] = pinSession;
    await writePinSessions(sessions);
  }
  
  return pinSession;
}

/**
 * 使用PIN码登录
 * @param {string} pin 
 * @returns {Promise<object|null>} { username } 或 null
 */
async function usePinCode(pin) {
  const sessions = await readPinSessions();
  const pinSession = sessions[pin];
  
  // PIN码不存在
  if (!pinSession) return null;
  
  // 已使用
  if (pinSession.status !== PINStatus.PENDING) return null;
  
  // 已过期
  if (Date.now() > pinSession.expiresAt) {
    pinSession.status = PINStatus.EXPIRED;
    sessions[pin] = pinSession;
    await writePinSessions(sessions);
    return null;
  }
  
  // 标记为已使用
  pinSession.status = PINStatus.USED;
  sessions[pin] = pinSession;
  await writePinSessions(sessions);
  
  return { username: pinSession.username };
}

/**
 * 删除PIN码会话
 * @param {string} pin 
 */
async function deletePinSession(pin) {
  const sessions = await readPinSessions();
  if (sessions[pin]) {
    delete sessions[pin];
    await writePinSessions(sessions);
  }
}

/**
 * 清理过期会话
 */
async function cleanExpiredPinSessions() {
  const sessions = await readPinSessions();
  const now = Date.now();
  let changed = false;
  
  for (const [pin, session] of Object.entries(sessions)) {
    if (now > session.expiresAt) {
      delete sessions[pin];
      changed = true;
    }
  }
  
  if (changed) {
    await writePinSessions(sessions);
  }
}

/**
 * 清除缓存
 */
function clearCache() {
  pinCache = null;
  cacheTime = 0;
}

// 每5分钟清理过期会话
setInterval(cleanExpiredPinSessions, 5 * 60 * 1000);

module.exports = {
  createPinSession,
  getPinSession,
  usePinCode,
  deletePinSession,
  cleanExpiredPinSessions,
  clearCache,
  PINStatus
};

File: go-impl/internal/model/pinCodeSession.go
package model

import (
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// PINStatus PIN码状态
type PINStatus string

const (
	PINStatusPending PINStatus = "pending"
	PINStatusUsed    PINStatus = "used"
	PINStatusExpired PINStatus = "expired"
)

// PINSession PIN码会话
type PINSession struct {
	PIN       string    `json:"pin"`
	Username  string    `json:"username"`
	Status    PINStatus `json:"status"`
	CreatedAt int64     `json:"createdAt"`
	ExpiresAt int64     `json:"expiresAt"`
}

// PINCodeManager PIN码管理器
type PINCodeManager struct {
	filePath string
	sessions sync.Map // pin -> *PINSession
	expiry   time.Duration
	dirty    chan struct{}
}

var (
	pinCodeManager *PINCodeManager
	pinOnce        sync.Once
)

// GetPINCodeManager 获取PIN码管理器单例
func GetPINCodeManager() *PINCodeManager {
	pinOnce.Do(func() {
		execPath, _ := os.Executable()
		baseDir := filepath.Dir(execPath)
		pinCodeManager = &PINCodeManager{
			filePath: filepath.Join(baseDir, "data", "pinSessions.json"),
			expiry:   5 * time.Minute,
			dirty:    make(chan struct{}, 1),
		}
		pinCodeManager.load()
		go pinCodeManager.startCleanup()
		go pinCodeManager.autoSave()
	})
	return pinCodeManager
}

func (pm *PINCodeManager) load() {
	data, err := os.ReadFile(pm.filePath)
	if err != nil {
		return
	}
	var sessions map[string]*PINSession
	if json.Unmarshal(data, &sessions) == nil {
		for pin, s := range sessions {
			pm.sessions.Store(pin, s)
		}
	}
}

func (pm *PINCodeManager) save() {
	sessions := make(map[string]*PINSession)
	pm.sessions.Range(func(key, value interface{}) bool {
		sessions[key.(string)] = value.(*PINSession)
		return true
	})

	os.MkdirAll(filepath.Dir(pm.filePath), 0755)
	data, _ := json.MarshalIndent(sessions, "", "  ")
	os.WriteFile(pm.filePath, data, 0644)
}

func (pm *PINCodeManager) autoSave() {
	for range pm.dirty {
		pm.save()
	}
}

func (pm *PINCodeManager) markDirty() {
	select {
	case pm.dirty <- struct{}{}:
	default:
	}
}

// generatePIN 生成4位数字PIN码
func generatePIN() string {
	return fmt.Sprintf("%04d", 1000+rand.Intn(9000))
}

// CreatePINSession 创建PIN码会话
func (pm *PINCodeManager) CreatePINSession(username string) (string, int64) {
	// 删除该用户之前未使用的PIN码
	pm.sessions.Range(func(key, value interface{}) bool {
		session := value.(*PINSession)
		if session.Username == username && session.Status == PINStatusPending {
			pm.sessions.Delete(key)
		}
		return true
	})

	// 生成唯一PIN码
	pin := generatePIN()
	for {
		if _, ok := pm.sessions.Load(pin); !ok {
			break
		}
		pin = generatePIN()
	}

	now := time.Now()
	session := &PINSession{
		PIN:       pin,
		Username:  username,
		Status:    PINStatusPending,
		CreatedAt: now.UnixMilli(),
		ExpiresAt: now.Add(pm.expiry).UnixMilli(),
	}

	pm.sessions.Store(pin, session)
	pm.markDirty()

	return pin, session.ExpiresAt
}

// GetPINSession 获取PIN码会话
func (pm *PINCodeManager) GetPINSession(pin string) *PINSession {
	value, ok := pm.sessions.Load(pin)
	if !ok {
		return nil
	}

	session := value.(*PINSession)
	if time.Now().UnixMilli() > session.ExpiresAt {
		session.Status = PINStatusExpired
	}

	return session
}

// UsePINCode 使用PIN码
func (pm *PINCodeManager) UsePINCode(pin string) *string {
	value, ok := pm.sessions.Load(pin)
	if !ok {
		return nil
	}

	session := value.(*PINSession)
	if session.Status != PINStatusPending {
		return nil
	}

	if time.Now().UnixMilli() > session.ExpiresAt {
		session.Status = PINStatusExpired
		pm.markDirty()
		return nil
	}

	session.Status = PINStatusUsed
	pm.markDirty()

	return &session.Username
}

// DeletePINSession 删除PIN码会话
func (pm *PINCodeManager) DeletePINSession(pin string) {
	pm.sessions.Delete(pin)
	pm.markDirty()
}

func (pm *PINCodeManager) startCleanup() {
	ticker := time.NewTicker(5 * time.Minute)
	for range ticker.C {
		pm.cleanup()
	}
}

func (pm *PINCodeManager) cleanup() {
	now := time.Now().UnixMilli()
	pm.sessions.Range(func(key, value interface{}) bool {
		session := value.(*PINSession)
		if now > session.ExpiresAt {
			pm.sessions.Delete(key)
		}
		return true
	})
	pm.markDirty()
}
